src/modules/job-info/update.c

```diff
@@ -0,0 +1,720 @@
+        errno = EPROTO;
+        errmsg = "update-watch request rejected with invalid flag";
+        goto error;
+    }
+    if (!flux_msg_is_streaming (msg)) {
+        errno = EPROTO;
+        errmsg = "update-watch request rejected without streaming RPC flag";
+        goto error;
+    }
+    if (!streq (key, "R")) {
+        errno = EINVAL;
+        errmsg = "update-watch unsupported key specified";
+        goto error;
+    }
+
+    if (!(index_key = get_index_key (id, key)))
+        goto error;
+
+    /* if no watchers for this jobid yet, start it */
+    if (!(uc = zhashx_lookup (ctx->index_uw, index_key))) {
+        if (update_lookup (ctx,
+                           UPDATE_TYPE_WATCH,
+                           msg,
+                           id,
+                           key,
+                           flags) < 0)
+            goto error;
+    }
+    else {
+        if (uc->update_object) {
+            if (flux_msg_authorize (msg, uc->userid) < 0)
+                goto error;
+            if (flux_respond_pack (uc->ctx->h, msg, "{s:O}",
+                                   uc->key, uc->update_object) < 0) {
+                flux_log_error (ctx->h, "%s: flux_respond", __FUNCTION__);
+                goto error;
+            }
+        }
+        /* if uc->update_object has not been set, the initial lookup
+         * has not completed.  The security check will be done in
+         * watch_lookup_continuation when the initial lookup completes.
+         */
+        flux_msglist_append (uc->msglist, msg);
+    }
+
+    free (index_key);
+    return;
+
+error:
+    if (flux_respond_error (h, msg, errno, errmsg) < 0)
+        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
+    free (index_key);
+}
+
+/* Cancel update_watch if it matches message.
+ */
+static void update_watch_cancel (struct update_ctx *uc,
+                                 const flux_msg_t *msg,
+                                 bool cancel)
+{
+    const flux_msg_t *cmpmsg;
+
+    cmpmsg = flux_msglist_first (uc->msglist);
+    while (cmpmsg) {
+        bool match;
+
+        if (cancel)
+            match = flux_cancel_match (msg, cmpmsg);
+        else
+            match = flux_disconnect_match (msg, cmpmsg);
+
+        if (match) {
+            if (flux_respond_error (uc->ctx->h, cmpmsg, ENODATA, NULL) < 0)
+                flux_log_error (uc->ctx->h,
+                                "%s: flux_respond_error",
+                                __FUNCTION__);
+            /* deletes at cursor */
+            flux_msglist_delete (uc->msglist);
+        }
+
+        cmpmsg = flux_msglist_next (uc->msglist);
+    }
+
+    if (flux_msglist_count (uc->msglist) == 0)
+        eventlog_watch_cancel (uc);
+}
+
+void update_watchers_cancel (struct info_ctx *ctx, const flux_msg_t *msg, bool cancel)
+{
+    struct update_ctx *uc;
+
+    uc = zlist_first (ctx->update_watchers);
+    while (uc) {
+        update_watch_cancel (uc, msg, cancel);
```

Comment:
Maybe it would be better to replace `update_watch_cancel()` with
```c
if (cancel)
    flux_msglist_cancel (ctx->h, uc->msglist, msg);
else
    flux_msglist_disconnect (ctx->h, uc->msglist, msg);
```